libs/types/src/locations.rs

```diff
@@ -33,21 +33,24 @@ pub enum Location {
 	Address(DomainAddress),
 }
 
-impl From<AccountId32> for Location {
+impl From<AccountId32> for RestrictedTransferLocation {
 	fn from(value: AccountId32) -> Self {
 		Self::Local(value)
 	}
 }
 
-impl From<VersionedMultiLocation> for Location {
-	fn from(vml: VersionedMultiLocation) -> Self {
+impl From<VersionedLocation> for RestrictedTransferLocation {
+	fn from(vml: VersionedLocation) -> Self {
 		// using hash here as multilocation is significantly larger than any other enum
 		// type here -- 592 bytes, vs 40 bytes for domain address (next largest)
 		Self::XCM(BlakeTwo256::hash(&vml.encode()))
+
+		// TODO-1.7: I'm afraid of locations translated from v3 to v4 will generate a different hash here.
```

Comment:
TODO here

Reply:
We're storing the `RestrictedTransferLocation` in `pallet-transfer-allowlist` so changing versions here will affect that storage

Reply:
We need to ensure there is no `RestrictedTransferLocation::Xcm` stored here. If there are, we will need to tackle this problem.

Reply:
It could not be an issue if the person who called `add_transfer_allowance(location.v3)` calls `remove_transfer_allowance(location.v3)` with the same version.

Reply:
Unfortunately, there is one XCM location in the `AccountCurrencyTransferAllowance` storage:
```
key ["4dTeMxuPJCK7zQGhFcgCivSJqBs9Wo2SuMSQeYCCuVJ9xrE2",{"Specific":{"ForeignAsset":"6"}},{"XCM":"0x8938623071e6f6b3eebd4ae4587d3e71b9e94aa00e7cfdac883829cd003aa23f"}]
val {"allowedAt":"0","blockedAt":"4,294,967,295"}

key ["4dTeMxuPJCK7zQGhFcgCivSJqBs9Wo2SuMSQeYCCuVJ9xrE2","All",{"Address":{"EVM":["1","0x1a3b234fb570f93eb2244796f0527f834cd30294"]}}]
val {"allowedAt":"0","blockedAt":"4,294,967,295"}

key ["4btLaQXkQRkV6SfHk5VbHvb4Dzipbg4r2hRCfrDJYYiKCfHW",{"Specific":{"ForeignAsset":"100,001"}},{"Address":{"EVM":["1","0x1a3b234fb570f93eb2244796f0527f834cd30294"]}}]
val {"allowedAt":"0","blockedAt":"4,294,967,295"}
```

Since we cannot guarantee that new ones won't be added in between proposing and executing the RU, I suppose I should add a migration for this? I.e. decoding the V3 Location from the stored hash and storing as V4 hash.

Another solution would be adding this translation layer to the `From<VersionedLocation>` impl of `RestrictedTransferLocation`.q

Reply:
> decoding the V3 Location from the stored hash and storing as V4 hash.

And how you can know what was the initial location v3 struct? Not sure if we store these locations somewhere to rebuild them, translate them to v4 and make the hash again.

> Another solution would be adding this translation layer to the `From<VersionedLocation>` impl of `RestrictedTransferLocation`.

Could you expand?

Reply:
> We need to set a max size then as locations can be arbitrarily large.

Good point to think about! 
In which sense it can be "arbitrarily"? It can have 9 elements as much (given by `X9`), and looking deeply they can not be nested

Reply:
> so we can definitely due this manually for this case.

I guess we'll go with this option, right?

Reply:
Or using directly a `Location` there. What do you prefer? I can do that

Reply:
> In which sense it can be "arbitrarily"? It can have 9 elements as much (given by `X9`), and looking deeply they can not be nested

Got me with my superficial knowledge 😅 Suppose then it's fine because it is bound by the `MaxEncodedLen` which will of course be a heavy overestimation.

Reply:
Will do then!

Reply:
https://github.com/centrifuge/centrifuge-chain/pull/1853 and https://github.com/centrifuge/centrifuge-chain/pull/1858